File: everything_parsing/16-lexer-project1/js/subset.mjs
// Module 13 — the subset (powerset) construction: NFA -> equivalent DFA.
//
// A DFA state is a SET of NFA states — "every state a clone could be in."
//   DFA start        = epsilonClosure({ nfa.start })
//   DFA delta(S, a)  = epsilonClosure( union of nfa.move(s, a) for s in S )
//   DFA state S accepting  iff  S meets nfa.accept
//
// Only reachable subsets are built (worklist), so the DFA is usually much
// smaller than 2^|NFA states| — but the worst case is exactly that big.

import { Nfa } from "./nfa.mjs";
import { Dfa } from "./dfa.mjs";

const key = (set) => [...set].sort().join("");

function explore(nfa) {
  // returns { order: [ [name, sortedSet] ], deltaByName: Map, acceptNames: [] }
  const nameOf = new Map();               // key -> name
  const setOf = new Map();                // name -> sorted array
  const order = [];
  const work = [];

  const start = [...nfa.epsilonClosure(new Set([nfa.start]))].sort();
  nameOf.set(key(start), "D0");
  setOf.set("D0", start);
  order.push("D0");
  work.push(start);

  const deltaByName = new Map();
  const acceptNames = [];

  while (work.length) {
    const S = work.shift();
    const sName = nameOf.get(key(S));
    if (S.some(s => nfa.accept.has(s))) acceptNames.push(sName);

    for (const a of nfa.alphabet) {
      const moved = new Set();
      for (const s of S) for (const t of nfa.move(s, a)) moved.add(t);
      const T = [...nfa.epsilonClosure(moved)].sort();
      const k = key(T);
      let tName = nameOf.get(k);
      if (tName === undefined) {
        tName = "D" + nameOf.size;
        nameOf.set(k, tName);
        setOf.set(tName, T);
        order.push(tName);
        work.push(T);
      }
      if (!deltaByName.has(sName)) deltaByName.set(sName, new Map());
      deltaByName.get(sName).set(a, tName);
    }
  }
  return { order, setOf, deltaByName, acceptNames };
}

export function determinize(nfa) {
  const { order, deltaByName, acceptNames } = explore(nfa);
  let text = "states: " + order.join(" ") + "\n";
  text += "alphabet: " + nfa.alphabet.join(" ") + "\n";
  text += "start: D0\n";
  text += "accept: " + acceptNames.join(" ") + "\n";
  for (const [from, m] of deltaByName)
    for (const [a, to] of m)
      text += from + " " + a + " " + to + "\n";
  return Dfa.parse(text);
}

export function legend(nfa) {
  const { order, setOf } = explore(nfa);
  const out = new Map();
  for (const n of order) out.set(n, setOf.get(n));
  return out;
}

// NFA over {a,b} for "the k-th symbol from the end is 'a'": k+1 states, DFA 2^k.
export function kthFromEndNfa(k) {
  let text = "states:";
  for (let i = 0; i <= k; i++) text += " s" + i;
  text += "\nalphabet: a b\nstart: s0\naccept: s" + k + "\n";
  text += "s0 a s0\ns0 b s0\ns0 a s1\n";
  for (let i = 1; i < k; i++) {
    text += "s" + i + " a s" + (i + 1) + "\n";
    text += "s" + i + " b s" + (i + 1) + "\n";
  }
  return Nfa.parse(text);
}

File: everything_parsing/15-maximal-munch-priority/java/Subset.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;

/**
 * Module 13 — the subset (powerset) construction: turn an NFA into an equivalent
 * DFA.
 *
 * A DFA state is a SET of NFA states — "every state a clone could be in."
 *
 *   DFA start   = epsilonClosure({ nfa.start })
 *   DFA delta(S, a) = epsilonClosure( union of nfa.move(s, a) for s in S )
 *   DFA state S is accepting  iff  S contains an NFA accepting state
 *
 * Only the *reachable* subsets are built (worklist), so the DFA is often far
 * smaller than 2^|NFA states| — but in the worst case it is exactly that big,
 * which `kthFromEndNfa` demonstrates.
 */
public final class Subset {
    private Subset() {}

    public static Dfa determinize(Nfa nfa) {
        // reachable subsets, in discovery order; each gets a name D0, D1, ...
        Map<Set<String>, String> name = new LinkedHashMap<>();
        Deque<Set<String>> work = new ArrayDeque<>();

        Set<String> startSet = nfa.epsilonClosure(Set.of(nfa.start));
        name.put(startSet, "D0");
        work.add(startSet);

        Map<String, Map<String, String>> delta = new LinkedHashMap<>();
        List<String> accepting = new ArrayList<>();

        while (!work.isEmpty()) {
            Set<String> S = work.poll();
            String sName = name.get(S);
            if (!Nfa.intersect(S, nfa.accept).isEmpty()) accepting.add(sName);

            for (String a : nfa.alphabet) {
                Set<String> moved = new TreeSet<>();
                for (String s : S) moved.addAll(nfa.move(s, a));
                Set<String> T = new TreeSet<>(nfa.epsilonClosure(moved));
                String tName = name.get(T);
                if (tName == null) {
                    tName = "D" + name.size();
                    name.put(T, tName);
                    work.add(T);
                }
                delta.computeIfAbsent(sName, k -> new LinkedHashMap<>()).put(a, tName);
            }
        }

        // emit .dfa text and reuse Module 10's loader
        StringBuilder sb = new StringBuilder("states:");
        for (String n : name.values()) sb.append(" ").append(n);
        sb.append("\nalphabet:");
        for (String a : nfa.alphabet) sb.append(" ").append(a);
        sb.append("\nstart: D0");
        sb.append("\naccept:");
        for (String n : accepting) sb.append(" ").append(n);
        sb.append("\n");
        for (var e : delta.entrySet())
            for (var t : e.getValue().entrySet())
                sb.append(e.getKey()).append(" ").append(t.getKey()).append(" ").append(t.getValue()).append("\n");
        return Dfa.parse(sb.toString());
    }

    /** The legend: which set of NFA states each DFA state name stands for. */
    public static Map<String, Set<String>> legend(Nfa nfa) {
        Map<Set<String>, String> name = new LinkedHashMap<>();
        Deque<Set<String>> work = new ArrayDeque<>();
        Set<String> startSet = nfa.epsilonClosure(Set.of(nfa.start));
        name.put(startSet, "D0");
        work.add(startSet);
        while (!work.isEmpty()) {
            Set<String> S = work.poll();
            for (String a : nfa.alphabet) {
                Set<String> moved = new TreeSet<>();
                for (String s : S) moved.addAll(nfa.move(s, a));
                Set<String> T = new TreeSet<>(nfa.epsilonClosure(moved));
                if (!name.containsKey(T)) { name.put(T, "D" + name.size()); work.add(T); }
            }
        }
        Map<String, Set<String>> out = new LinkedHashMap<>();
        for (var e : name.entrySet()) out.put(e.getValue(), e.getKey());
        return out;
    }

    /**
     * An NFA over {a,b} accepting "the k-th symbol from the end is a".
     * It has k+1 states; its minimal DFA has 2^k.
     */
    public static Nfa kthFromEndNfa(int k) {
        StringBuilder sb = new StringBuilder("states:");
        for (int i = 0; i <= k; i++) sb.append(" s").append(i);
        sb.append("\nalphabet: a b\nstart: s0\naccept: s").append(k).append("\n");
        sb.append("s0 a s0\ns0 b s0\n");          // loop on anything
        sb.append("s0 a s1\n");                    // guess: this 'a' is k-th from the end
        for (int i = 1; i < k; i++) {
            sb.append("s").append(i).append(" a s").append(i + 1).append("\n");
            sb.append("s").append(i).append(" b s").append(i + 1).append("\n");
        }
        return Nfa.parse(sb.toString());
    }
}
